### PR TITLE
🚀 build assets with webpack - publish to npm - install to octofarm

### PR DESCRIPTION
**Pull request by davidzwa**
NEW UI BUNDLING AND DISTRIBUTION

We download the `@octofarm/client` package from NPM and push to NPM regularly
https://www.npmjs.com/package/@octofarm/client

PR merged => webpack build + npm publish => npm new nightly version+build number (TODO)
release => webpack build + npm publish => npm new release version

- [x] jquery $ missing after server down/amialive detector
- [x] babel polyfill loading twice
- [x] ~Babel polyfill still working? (test)~
- [x] ~onComplete error (settings page)~ error not relevant to webpack
- [x] Bootbox does not work on printermanagement (had to fix with CDN include :/)
- [x] Fails on drag and drop (dashboard)
- [x] Does not place dashboard tiles properly (gridstack HTML5 implementation is kaputt, jQuery version wont work at all.)
- [x] Rewrite dashboard ejs, runner and worker
- [x] Add webpack workflow
- [x] TODO's with series data
- [x] Weekly Activity HeatMap legend markers were sagging
- [x] graph time not properly formatted as time
- [x] Print Completion By Day was missing title bar
- [x] SSE to apex error after updateChartSeries()
- [x] Changelog
- [x] Add push to NPM workflow if release/tag
- [x] File Manager page bug 
- [x] Filament Manager page bug 
- [x] bug in dashboard Unix (file case sensitivity grid-stack-item partial)
- [x] snapshots typo in system
    
Some stuff to note:
- weekly util only one day loaded (probably this is normal behaviour when the server is restarted frequently?)
- change event after page rescale causes save, but the layout is off and thus saved completely wrong

### CODE REVIEW

**Comment by davidzwa** [comment]
@NotExpectedYet can you test this and see how to fix the mentioned problems? There will be a lot more still, but they seem to work fine.
After this comes css... maybe I chickened out a bit too soon last time

**Comment by NotExpectedYet** [comment]
Yeah no worries man will take a look after work tomorrow

**Comment by davidzwa** [comment]
image
When reloading the server this error threw. Luckily this seems to be fixable! Now... lets get to it

**Comment by davidzwa** [comment]
image

Also gotta fix this. Something on dashboard is loading the polyfill twice.

**Comment by NotExpectedYet** [comment]
Weird the polyfil is commented out on the template so not sure on that one. 

Yeah I'd love to nuke jQuery. Wonder how hard it would be to upgrade to the new bootstrap that doesn't use it. There may be other depencies reliant on it too which makes it somewhat problematic.

**Comment by NotExpectedYet** [comment]
The amIalive check only uses it for the modal pop up wonder if we could change that to pure JS

**Comment by davidzwa** [comment]
I think we can fix the jquery missing error. To go for a new bootstrap is gonna be a huge effort, as well as replacing it with Noty. Let's first hack away on the problem a bit. Shouldn't be too hard.

**Comment by NotExpectedYet** [comment]
> I think we can fix the jquery missing error. To go for a new bootstrap is gonna be a huge effort, as well as replacing it with Noty. Let's first hack away on the problem a bit. Shouldn't be too hard.
> 
> **Polyfill double issue**
> Well I can guess that loading the bundle as well as the custom bundles might cause that. There's a flag in babel to disable the warning. The only thing they say is: 'dont mix versions and everything is still good' and tbh I can be content with that idea if it makes my life easier.

I wasn't meaning replace to the new bootstrap, still in beta. Was just meaning don't use the modal function on the iamalive, we could easily include the UI file for the create alert as an alertnative there. 

**Poyfill**
I'm unsure how webpack loads it, gather it's similar to gulp. I may take a sneak at this when things wind down on Monday.

**Comment by davidzwa** [comment]
- Ive found that removing `import "@babel/polyfill";` on the top of f.e. client printermanagementrunner.js fixes the problem.
- Jquery error is fixed.

So now 'only' the gridstack drag n drop.

**Comment by davidzwa** [comment]
chokidar dep for @babel/cli was being a problem. Babel changed it from 7.13.0 to 7.13.10 and made it optional. That was the fix for us (I ran `npm update` which updated the lock fully).

**Comment by NotExpectedYet** [comment]
Are the last two done on this? 

CAF btw.

**Comment by davidzwa** [comment]
Latest state
image

Dashboard code's a mess.

**Comment by NotExpectedYet** [comment]
How does one build with this? I tried with just `webpack` and got Noty errors.

**Comment by davidzwa** [comment]
Run webpack --watch

**Comment by davidzwa** [comment]
Ofc requires the npm packages to be up to date

**Comment by NotExpectedYet** [comment]
The error is below, feel it's me been dumb as I can see node_modules/noty, It's in package.json although needs updating as the beta version isn't supported anymore. 

I've run:
`npm ci`
`npm install`
`npm install noty`

All still had the error with `webpack` and `webpack --watch`

``ERROR in ./client_src/js/lib/functions/ui.js 50:4-8
Module not found: Error: Can't resolve 'Noty' in '/home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js/lib/functions'
resolve 'Noty' in '/home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js/lib/functions'
  Parsed request is a module
  using description file: /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/package.json (relative path: ./client_src/js/lib/functions)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js/lib/functions/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js/lib/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/node_modules doesn't exist or is not a directory
      looking for modules in /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules
        single file module
          using description file: /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/package.json (relative path: ./node_modules/Noty)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty.js doesn't exist
            .json
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty.json doesn't exist
            .wasm
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty.wasm doesn't exist
        /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty doesn't exist
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/NotExpectedYet Development/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/node_modules doesn't exist or is not a directory
      /home/s1mpleman/node_modules doesn't exist or is not a directory
      /home/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory
 @ ./client_src/js/currentOperationsViewRunner.js 3:0-39 18:6-20

ERROR in ./client_src/js/serverAliveCheck.js 38:18-22
Module not found: Error: Can't resolve 'Noty' in '/home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js'
resolve 'Noty' in '/home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js'
  Parsed request is a module
  using description file: /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/package.json (relative path: ./client_src/js)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/js/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/client_src/node_modules doesn't exist or is not a directory
      looking for modules in /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules
        single file module
          using description file: /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/package.json (relative path: ./node_modules/Noty)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty.js doesn't exist
            .json
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty.json doesn't exist
            .wasm
              Field 'browser' doesn't contain a valid alias configuration
              /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty.wasm doesn't exist
        /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/366/node_modules/Noty doesn't exist
      /home/s1mpleman/Development/NotExpectedYet Development/octofarm-development/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/NotExpectedYet Development/node_modules doesn't exist or is not a directory
      /home/s1mpleman/Development/node_modules doesn't exist or is not a directory
      /home/s1mpleman/node_modules doesn't exist or is not a directory
      /home/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory

webpack 5.28.0 compiled with 2 errors in 5195 ms
``

**Comment by davidzwa** [comment]
@NotExpectedYet I cannot reproduce, even when running on both Linux and Windows.

My Ubuntu in WSL:
npm: `7.5.3`
node: `v15.10.0` 
image

On Windows:
npm: `7.8.0`
node: `v14.16.0` 
image

I did update the lock file today, as it was massively outdated. Please try again and also make sure your NPM, Node and node_modules are 100% up to date.

**Comment by davidzwa** [comment]
Also note that when I run npm install to lock the dependencies, I currently produce lockFileVersion: 2 in the produced lock-file.

```
{ 
    ...
    "lockfileVersion": 2,
    ...
}
```
Not sure if that matters, but it could explain the massive change in lock file.

**Comment by NotExpectedYet** [comment]
Sound I'll have a look

**Comment by davidzwa** [comment]
Currently the tiles at least load and allow placement. But dragging and dropping has a certain offset and also the height for each widget saved is huge. 

So the tiles start out ok-ish (like the height is a bit low, fixable with tweaking), but end up taking enormous height after a page load. I guess some height unit got a breaking change.

**Comment by NotExpectedYet** [comment]
> 
> 
> Currently the tiles at least load and allow placement in horizontal direction (this was broken). But dragging and dropping has a certain offset (not even predictable) and also the height for each widget saved is huge.
> 
> So the tiles start out ok-ish (like the height is a bit low, fixable with tweaking), but end up taking enormous height after a page load. I guess some height unit got a breaking change.
> 
> So 2 problems still, but getting closer and closer!

Have you tried with disabling the localStorage for this? Wondering if the options have changed there.

**Comment by davidzwa** [comment]
image
FINALLY FIXED IT!!!! Webpack is gonna be a go soon 😃

**Comment by davidzwa** [comment]
image
Something is off right now and it has to do with the top/left/right/bottom (inset) properties... I have no clue what is happening... again

**Comment by davidzwa** [comment]
It was `slate.css` doing a height: 100%!important .... great

**Comment by davidzwa** [comment]
image
Its nearly working (days not loading properly and the `dashboard.options.js` file has some dupes/TODO's missing)

File: babel.config.js
Comment (davidzwa) [inline babel.config.js:13]: Test

File: client_src/js/dashboard.runner.js
Comment (davidzwa) [inline client_src/js/dashboard.runner.js:63]: TODO

File: webpack/webpack.watcher.js
Comment (davidzwa) [inline webpack/webpack.watcher.js:3]: Fix this to be a dev feature

File: webpack.config.js
Comment (davidzwa) [inline webpack.config.js:45]: Fix dev/prod mode

File: webpack.config.js
Comment (davidzwa) [inline webpack.config.js:53]: Fix dev/prod

File: client_src/js/dashboard/dashboard.options.js
Comment (davidzwa) [inline client_src/js/dashboard/dashboard.options.js:413]: TODO series name

File: client_src/js/dashboard/dashboard.options.js
Comment (davidzwa) [inline client_src/js/dashboard/dashboard.options.js:739]: TODO availableStats

File: client_src/js/dashboard/dashboard.options.js
Comment (davidzwa) [inline client_src/js/dashboard/dashboard.options.js:496]: TODO yAxis

File: client_src/js/dashboard/dashboard.options.js
Comment (davidzwa) [inline client_src/js/dashboard/dashboard.options.js:512]: TODO Dupe

**Comment by davidzwa** [comment]
image
- icon sagging
- graph time not properly formatted as time

**Comment by davidzwa** [comment]
@NotExpectedYet this PR is done. The failing workflow should fix itself and otherwise I will follow up with a fix or a disable.
The PR cannot be merged in 1.2 as it is a major update.
`1.2.1` would be nice but not entirely correct. I dont want to drag this PR along any further, so that's why I propose 1.2.1

**Comment by NotExpectedYet** [comment]
@davidzwa Ok will make some time and test tomorrow. 

I'm happy for it to go into 1.2. are we holding off on the RC before that then?

**Comment by davidzwa** [comment]
Happy to wait for the RC if it allows us to include webpack.

**Comment by NotExpectedYet** [comment]
I think it should personally. Also gives me time to switch the website over.

**Comment by NotExpectedYet** [comment]
Could I get a little run down of how this works now? 

Like my only experience with webpack is with vueJS and I probably gather I'm crossing wires. 

Does webpack run its own server when we do the development build? Or is it unpacking into the directory OF serves from. 

What about changes going forward to the client? I gather we won't be building before push anymore?

**Comment by NotExpectedYet** [comment]
Ok checked the flows, my first question seems to be the latter, it's unpacked and same port as OF server. 

Second one seems it's automated looking at the GitHub actions :)

**Comment by davidzwa** [comment]
Yes and good point I should add that when NODE_ENV is equal to production and the NPM package is installed it will serve the NPM package dist folder instead.
I think this mechanism is super bug prone as the flag must be production. Gotta think about making that way more robust.
Stuff to think about

**Comment by NotExpectedYet** [comment]
I noticed the weekly util, and I think that is the case btw. It depends when it saves to the db, if it's straight away there is a bug and that's on current development too so not contained to this webpack branch. Probably from one of the cleanups.

**Comment by davidzwa** [comment]
We will fix it then. Good! Funny how a bug not-on-my-branch makes me feel its not for me to fix. Ofc not true

**Comment by NotExpectedYet** [comment]
@davidzwa I tend not to play the blame game, a bug is a bug regardless of who caused it. It's a big app, getting bigger shits bound to get crossed and explode in a glory of bug guts. 
I need to look at the code to refresh my head anyway before I'll commit to it been a bug. I think (if I remember rightly) I save that state to the dB every few hours. It's working on the last development instance I had on my Pi4 which was after the cleaner changes so it's probably fine. I don't think it needs to save more than that either. It is a server application after all.

**Comment by davidzwa** [comment]
@NotExpectedYet got some time for this? Wanna build stuff on top and test it before RC to be frank.
I know stuff will be breaking :)

**Review by NotExpectedYet — APPROVED** [review]

**Comment by NotExpectedYet** [comment]
@davidzwa before you merge just change the changelog title back to rc1 seen as it got moved into that.

File: app-core.js
Comment (davidzwa) [inline app-core.js:30]: So we have to expand this logic later. We could offer a direct bundle download without NPM as backup, or use a github release asset. Whatever we feel like, but it needs more user interactivity.

I noted this in my mind and expect to enhance it later! Oh btw remember the `NODE_ENV` also needs to be in production by default, made sure to be correct... whatever is most robust.

**Pull request merged**
